app/models/sso/identity.rb

```diff
@@ -0,0 +1,62 @@
+module SSO
+  class Identity
+    InvalidData = Class.new(StandardError)
+
+    class << self
+      def from_omniauth(auth_hash)
+        info = auth_hash.fetch('info')
+        user = User.from_omniauth(auth_hash)
+        options = extract_additional_info(info)
+        new(user, options)
+      rescue
+        nil
+      end
+
+      def from_session(hash)
+        user = User.find(hash.fetch('user_id'))
+        options = {
+          profile_url: hash.fetch('profile_url'),
+          logout_url: hash.fetch('logout_url'),
+          permissions: hash.fetch('permissions')
+        }
+        new(user, options)
+      rescue KeyError
```

Comment:
Don't think this is necessary. 🛡 ⚔️ 

It would suggest we've allowed bad data into the system to begin with(as its session retrieval).

For the `options` we could replace the fetches with a `slice`.